src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -367,24 +394,128 @@ private static String getClassNameFromQualifiedName(String qualifiedName) {
     }
 
-    /** Maintains information about class' ctors. */
-    private static final class ClassDesc {
+    /**
+     * Calculates and returns the type declaration matching count when {@code classToBeMatched} is
+     * considered to be super class of an anonymous inner class.
+     *
+     * <p>
+     * Suppose our pattern class is {@code Main.ClassOne} and class to be matched is
+     * {@code Main.ClassOne.ClassTwo.ClassThree} then type declaration name matching count would
+     * be calculated by comparing every character, and updating main counter when we hit "." or
+     * when it is the last character of the pattern class and certain conditions are met. This is
+     * done so that matching count is 13 instead of 5. This is due to the fact that pattern class
+     * can contain anonymous inner class object of a nested class which isn't true in case of
+     * extending classes as you can't extend nested classes.
+     * </p>
+     *
+     * @param patternTypeDeclaration type declaration against which the given type declaration has
+     *                               to be matched
+     * @param typeDeclarationToBeMatched type declaration to be matched
+     * @return type declaration matching count
+     */
+    private static int getAnonSuperTypeMatchingCount(String patternTypeDeclaration,
+                                                    String typeDeclarationToBeMatched) {
+        final int minLength = Math
+            .min(typeDeclarationToBeMatched.length(), patternTypeDeclaration.length());
+        final char packageSeparator = PACKAGE_SEPARATOR.charAt(0);
+        int result = 0;
+        final boolean shouldCountBeUpdatedAtLastCharacter =
+            typeDeclarationToBeMatched.length() > minLength
+                && typeDeclarationToBeMatched.charAt(minLength) == packageSeparator;
+
+        for (int index = 0; index < minLength
+            && patternTypeDeclaration.charAt(index) == typeDeclarationToBeMatched.charAt(index);
+             ++index) {
+            if (index == minLength - 1
+                && shouldCountBeUpdatedAtLastCharacter
+                || patternTypeDeclaration.charAt(index) == packageSeparator) {
+                result = index;
+            }
+        }
+        return result;
+    }
 
-        /** Corresponding node. */
-        private final DetailAST classAst;
+    /**
+     * Maintains information about the type of declaration.
+     * Any ast node of type {@link TokenTypes#CLASS_DEF} or {@link TokenTypes#INTERFACE_DEF}
+     * or {@link TokenTypes#ENUM_DEF} or {@link TokenTypes#ANNOTATION_DEF}
+     * or {@link TokenTypes#RECORD_DEF} is considered as a type declaration.
+     */
+    private static class TypeDeclarationDescription {
 
-        /** Qualified class name(with package). */
+        /**
+         * Complete type declaration name with package name and outer type declaration name.
+         */
         private final String qualifiedName;
```

Comment:
just in case - for private static classes there is no need for getters, you can access its fields directly

Reply:
Yes, we can, I created getters for using method reference here-
```
.max(Comparator
    .comparingInt(countProvider::apply)
    .thenComparingInt(ClassDesc::getDepth)) // <--- Over here
```
Now I created a getter for one field and it would look weird if I didn't create it for the others.

access modifier of getters can be default but XdocsJavaDocsTest doesn't allow it (can be a bug or a feature idk)